agent/agent.go

```diff
@@ -1226,7 +1226,7 @@ func (a *Agent) ShutdownEndpoints() {
 	a.shutdownLock.Lock()
 	defer a.shutdownLock.Unlock()
 
-	if len(a.dnsServers) == 0 || len(a.httpServers) == 0 {
+	if len(a.dnsServers) == 0 && len(a.httpServers) == 0 {
```

Comment:
This is intentional behavior added in https://github.com/hashicorp/consul/commit/ea5b0f2c7c9691a05699f902258dd1867463a89b, see comment for more context. We can return early if either one of dns/httpservers is empty because that means another `shutDownEndpoints` call is in progress.

Reply:
@preetapan thanks for your explanation.

i think `func (a *Agent) ShutdownEndpoints()` is protected by `a.shutdownLock`, if we are to proctect the race you mentioned, then the condition `either one of dns/httpservers is empty` should never happen?

Reply:
you are welcome :)